File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/authorization/SemanticAuthorizer.java
package com.foundgine.core.semantic.authorization;

import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.*;
import com.foundgine.core.semantic.ir.*;
import com.foundgine.core.semantic.ir.graph.*;
import java.util.*;

/** Applies authorization to resolved semantic graphs/operations before planning. */
public final class SemanticAuthorizer {
 private final ISemanticAuthorizationPolicy policy;
 public SemanticAuthorizer(ISemanticAuthorizationPolicy policy){this.policy=Objects.requireNonNull(policy);}

 public SemanticGraph authorize(SemanticGraph graph){
  Objects.requireNonNull(graph); if(graph.nodes().isEmpty())return graph; var authorized=new SemanticGraph();authorized.setOptions(graph.options());var map=new HashMap<Integer,SemanticGraph.SemanticGraphNode>();
  for(var source:graph.nodes()){
   SemanticGraph.SemanticGraphNode parent=null;if(source.parentId()!=null){parent=map.get(source.parentId());if(parent==null)continue;}
   var fieldDecisions=source.fields().stream().distinct().map(f->new AbstractMap.SimpleEntry<>(f,policy.getFieldAccess(source.entityId(),f,AuthorizationOperation.READ))).toList();
   var fields=fieldDecisions.stream().filter(x->x.getValue().isAllowed()).map(Map.Entry::getKey).toList();
   var fieldAuth=SemanticAuthorizationCapabilityComposition.compose(fieldDecisions.stream().filter(x->x.getValue().isAllowed()).map(Map.Entry::getValue).toList());
   var authorization=AuthorizationDecision.combine(policy.getEntityAccess(source.entityId(),AuthorizationOperation.READ),predicateDecision(policy.getPredicate(source.entityId(),AuthorizationOperation.READ)));
   authorization=AuthorizationDecision.combine(authorization,fieldAuth); if(!authorization.isAllowed()){if(source.parentId()==null)throw new SemanticAuthorizationException("Access denied for entity '"+source.entityId()+"'.");continue;}
   if(source.viaRelationship()!=null){if(source.parentId()==null||graph.nodes().stream().noneMatch(n->n.id()==source.parentId()))throw new IllegalArgumentException("Graph node "+source.id()+" has relationship '"+source.viaRelationship()+"' but no valid parent.");var parentEntity=graph.nodes().stream().filter(n->n.id()==source.parentId()).findFirst().orElseThrow().entityId();var rd=policy.getRelationshipAccess(parentEntity,source.viaRelationship(),AuthorizationOperation.READ);if(!rd.isAllowed())continue;authorization=AuthorizationDecision.combine(authorization,rd);}
   if(source.authorization()!=null&&!source.authorization().equals(authorization.predicate()))authorization=AuthorizationDecision.combine(authorization,predicateDecision(source.authorization()));
   SemanticGraph.SemanticGraphNode node;if(source.parentId()==null)node=authorized.addRoot(source.entityId(),fields,authorization.predicate());else if(source.viaRelationship()!=null)node=authorized.add(source.entityId(),source.viaRelationship(),parent,fields,authorization.predicate());else if(source.viaConnection()!=null)node=authorized.addConnection(source.entityId(),source.viaConnection(),parent,fields,authorization.predicate());else throw new IllegalArgumentException("Graph node "+source.id()+" has a parent but no semantic edge.");map.put(source.id(),node);
  } return authorized;
 }
 public SemanticOperationGraph authorize(SemanticContractSnapshot c,SemanticOperationGraph g){return authorizeGraphWithEvidence(c,g).graph();}
 public SemanticOperationGraphAuthorizationResult authorizeGraphWithEvidence(SemanticContractSnapshot c,SemanticOperationGraph g){Objects.requireNonNull(c);Objects.requireNonNull(g);var result=authorizeWithEvidence(c,g.toOperation());return new SemanticOperationGraphAuthorizationResult(SemanticOperationGraph.create(result.operation()),result.evidence());}
 public SemanticOperation authorize(SemanticContractSnapshot c,SemanticOperation op){return authorizeWithEvidence(c,op).operation();}
 public SemanticAuthorizationResult authorizeWithEvidence(SemanticContractSnapshot c,SemanticOperation op){Objects.requireNonNull(c);Objects.requireNonNull(op);SemanticAuthorizationContractValidator.validate(c,op);var authorized=authorize(op);return new SemanticAuthorizationResult(authorized,SemanticAuthorizationEvidence.create(c,authorized));}
 public SemanticOperation authorize(SemanticOperation op){Objects.requireNonNull(op);var root=authorizeNode(op.root(),true);if(root==null)throw new SemanticAuthorizationException("Access denied for entity '"+op.root().entityId()+"'.");return new SemanticOperation(root);}
 private SemanticReadNode authorizeNode(SemanticReadNode n,boolean root){
  var effective=AuthorizationDecision.combine(policy.getEntityAccess(n.entityId(),AuthorizationOperation.READ),predicateDecision(policy.getPredicate(n.entityId(),AuthorizationOperation.READ)));if(!effective.isAllowed()){if(root)throw new SemanticAuthorizationException("Access denied for entity '"+n.entityId()+"'.");return null;}
  var fields=new ArrayList<FieldId>();for(var f:new LinkedHashSet<>(n.fields())){var d=policy.getFieldAccess(n.entityId(),f,AuthorizationOperation.READ);if(d.isAllowed()){fields.add(f);effective=AuthorizationDecision.combine(effective,d);}}
  if(n.authorization()!=null&&!n.authorization().equals(effective.predicate()))effective=AuthorizationDecision.combine(effective,predicateDecision(n.authorization()));if(!effective.isAllowed()){if(root)throw new SemanticAuthorizationException("Authorization constraints denied semantic node '"+n.id()+"'.");return null;}
  var children=new ArrayList<SemanticReadNode>();for(var child:n.children()){
   if(child.viaRelationship()!=null){var rd=policy.getRelationshipAccess(n.entityId(),child.viaRelationship(),AuthorizationOperation.READ);if(!rd.isAllowed())continue;var ca=authorizeNode(child,false);if(ca==null)continue;var combined=AuthorizationDecision.combine(rd,ca.authorization()==null?AuthorizationDecision.ALLOWED:predicateDecision(ca.authorization()));children.add(ca.withAuthorization(combined.predicate()));}
   else {var ca=authorizeNode(child,false);if(ca!=null)children.add(ca);}
  }
  return new SemanticReadNode(n.id(),n.entityId(),fields,n.viaRelationship(),n.viaConnection(),children,n.queryOptions(),effective.predicate(),n.requiredFields());
 }
 private static AuthorizationDecision predicateDecision(AuthorizationPredicate p){return p==null?AuthorizationDecision.ALLOWED:AuthorizationDecision.conditional(p);}
}

File: src/csharp/Foundgine.Core/Semantic/Authorization/SemanticAuthorizer.cs
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.IR;
using Foundgine.Core.Semantic.IR.Graph;

namespace Foundgine.Core.Semantic.Authorization;

/// <summary>
/// Applies authorization to an already-resolved semantic graph.
/// The result remains provider- and protocol-independent.
/// </summary>
public sealed class SemanticAuthorizer
{
    private readonly ISemanticAuthorizationPolicy _policy;

    public SemanticAuthorizer(ISemanticAuthorizationPolicy policy)
    {
        _policy = policy ?? throw new ArgumentNullException(nameof(policy));
    }

    public SemanticGraph Authorize(SemanticGraph graph)
    {
        ArgumentNullException.ThrowIfNull(graph);

        if (graph.Nodes.Count == 0)
            return graph;

        var authorized = new SemanticGraph();
        authorized.Options = graph.Options;
        var sourceToAuthorized = new Dictionary<int, SemanticGraphNode>();

        foreach (var sourceNode in graph.Nodes)
        {
            SemanticGraphNode? authorizedParent = null;
            if (sourceNode.ParentId is not null &&
                !sourceToAuthorized.TryGetValue(sourceNode.ParentId.Value, out authorizedParent))
            {
                // An ancestor was denied, so this node is unreachable.
                continue;
            }

            var fieldDecisions = sourceNode.Fields
                .Distinct()
                .Select(fieldId => (
                    FieldId: fieldId,
                    Decision: _policy.GetFieldAccess(
                        sourceNode.EntityId, fieldId, AuthorizationOperation.Read)))
                .ToArray();

            var fields = fieldDecisions
                .Where(x => x.Decision.IsAllowed)
                .Select(x => x.FieldId)
                .ToArray();

            var fieldAuthorization = SemanticAuthorizationCapabilityComposition.Compose(
                fieldDecisions.Where(x => x.Decision.IsAllowed).Select(x => x.Decision));

            var authorization = AuthorizationDecision.Combine(
                _policy.GetEntityAccess(sourceNode.EntityId, AuthorizationOperation.Read),
                AuthorizationDecisionFromPredicate(_policy.GetPredicate(
                    sourceNode.EntityId, AuthorizationOperation.Read)));
            authorization = AuthorizationDecision.Combine(authorization, fieldAuthorization);

            if (!authorization.IsAllowed)
            {
                if (sourceNode.ParentId is null)
                    throw new SemanticAuthorizationException(
                        $"Access denied for entity '{sourceNode.EntityId}'.");

                continue;
            }

            if (sourceNode.ViaRelationship is { } relationshipId)
            {
                if (sourceNode.ParentId is not { } parentId ||
                    !graph.Nodes.Any(node => node.Id == parentId))
                {
                    throw new InvalidOperationException(
                        $"Graph node {sourceNode.Id} has relationship '{relationshipId}' but no valid parent.");
                }

                var parentEntityId = graph.Nodes.Single(node => node.Id == parentId).EntityId;
                var relationshipDecision = _policy.GetRelationshipAccess(
                    parentEntityId, relationshipId, AuthorizationOperation.Read);
                if (!relationshipDecision.IsAllowed)
                {
                    // A denied relationship removes this node and its descendants.
                    continue;
                }

                authorization = AuthorizationDecision.Combine(authorization, relationshipDecision);
            }

            var sourcePredicate = sourceNode.Authorization;
            if (sourcePredicate is not null && authorization.Predicate != sourcePredicate)
            {
                authorization = AuthorizationDecision.Combine(
                    authorization,
                    AuthorizationDecisionFromPredicate(sourcePredicate));
            }

            var predicate = authorization.Predicate;
            SemanticGraphNode node;
            if (sourceNode.ParentId is null)
            {
                node = authorized.AddRoot(sourceNode.EntityId, fields, predicate);
            }
            else if (sourceNode.ViaRelationship is { } childRelationshipId)
            {
                node = authorized.Add(sourceNode.EntityId, childRelationshipId, authorizedParent, fields, predicate);
            }
            else if (sourceNode.ViaConnection is { } connectionId)
            {
                node = authorized.AddConnection(sourceNode.EntityId, connectionId, authorizedParent!, fields,
                    predicate);
            }
            else
            {
                throw new InvalidOperationException($"Graph node {sourceNode.Id} has a parent but no semantic edge.");
            }

            sourceToAuthorized[sourceNode.Id] = node;
        }

        return authorized;
    }


    /// <summary>
    /// Authorizes canonical Semantic IR directly. This is the authoritative
    /// semantic-to-planning authorization boundary: physical providers are
    /// deliberately not involved.
    /// </summary>
    /// <summary>
    /// Authorizes canonical Semantic IR against the trusted immutable contract.
    /// The snapshot is validated before policy evaluation so authorization can
    /// never silently reason about identities that are outside the contract
    /// used to resolve and plan the operation.
    /// </summary>
    /// <summary>
    /// Authorizes the complete canonical operation graph as one security object.
    /// Every reachable entity, field and relationship edge is evaluated before
    /// the graph is returned to planning. The provider is deliberately absent
    /// from this boundary.
    /// </summary>
    public SemanticOperationGraph Authorize(
        SemanticContractSnapshot contract,
        SemanticOperationGraph graph)
    {
        return AuthorizeGraphWithEvidence(contract, graph).Graph;
    }

    /// <summary>
    /// Authorizes the complete operation graph and returns evidence bound to the
    /// exact semantic contract and resulting authorized operation.
    /// </summary>
    public SemanticOperationGraphAuthorizationResult AuthorizeGraphWithEvidence(
        SemanticContractSnapshot contract,
        SemanticOperationGraph graph)
    {
        ArgumentNullException.ThrowIfNull(contract);
        ArgumentNullException.ThrowIfNull(graph);

        // Graph validation and contract validation happen before policy evaluation.
        // Converting back to canonical IR also prevents the graph authorizer from
        // becoming a second semantic representation with different security rules.
        var operation = graph.ToOperation();
        var authorized = AuthorizeWithEvidence(contract, operation);
        var authorizedGraph = SemanticOperationGraph.Create(authorized.Operation);

        return new SemanticOperationGraphAuthorizationResult(
            authorizedGraph,
            authorized.Evidence);
    }

    public SemanticOperation Authorize(
        SemanticContractSnapshot contract,
        SemanticOperation operation)
    {
        return AuthorizeWithEvidence(contract, operation).Operation;
    }

    /// <summary>
    /// Authorizes an operation and returns immutable evidence bound to the exact
    /// semantic contract used for the decision.
    /// </summary>
    public SemanticAuthorizationResult AuthorizeWithEvidence(
        SemanticContractSnapshot contract,
        SemanticOperation operation)
    {
        ArgumentNullException.ThrowIfNull(contract);
        ArgumentNullException.ThrowIfNull(operation);

        SemanticAuthorizationContractValidator.Validate(contract, operation);
        var authorized = Authorize(operation);
        var evidence = SemanticAuthorizationEvidence.Create(contract, authorized);
        return new SemanticAuthorizationResult(authorized, evidence);
    }

    public SemanticOperation Authorize(SemanticOperation operation)
    {
        ArgumentNullException.ThrowIfNull(operation);

        var root = AuthorizeNode(operation.Root, isRoot: true);
        if (root is null)
            throw new SemanticAuthorizationException(
                $"Access denied for entity '{operation.Root.EntityId}'.");

        return new SemanticOperation(root);
    }

    private SemanticReadNode? AuthorizeNode(SemanticReadNode node, bool isRoot)
    {
        var entity = _policy.GetEntityAccess(node.EntityId, AuthorizationOperation.Read);
        var predicate = _policy.GetPredicate(node.EntityId, AuthorizationOperation.Read);

        var effective = AuthorizationDecision.Combine(
            entity,
            predicate is null
                ? AuthorizationDecision.Allowed
                : AuthorizationDecision.Conditional(predicate));

        // A denied entity removes a child subtree; a denied root rejects the
        // operation. This mirrors graph authorization while keeping the
        // canonical Semantic IR as the authorization boundary.
        if (!effective.IsAllowed)
        {
            if (isRoot)
                throw new SemanticAuthorizationException(
                    $"Access denied for entity '{node.EntityId}'.");
            return null;
        }

        var fieldDecisions = node.Fields
            .Distinct()
            .Select(fieldId => (
                FieldId: fieldId,
                Decision: _policy.GetFieldAccess(
                    node.EntityId, fieldId, AuthorizationOperation.Read)))
            .ToArray();

        var fields = new List<FieldId>(fieldDecisions.Length);
        foreach (var field in fieldDecisions)
        {
            if (!field.Decision.IsAllowed)
                continue;

            fields.Add(field.FieldId);
            effective = AuthorizationDecision.Combine(effective, field.Decision);
        }

        if (node.Authorization is not null)
            effective = AuthorizationDecision.Combine(
                effective,
                AuthorizationDecision.Conditional(node.Authorization));

        if (!effective.IsAllowed)
        {
            if (isRoot)
                throw new SemanticAuthorizationException(
                    $"Authorization constraints denied semantic node '{node.Id}'.");
            return null;
        }

        var children = new List<SemanticReadNode>();
        foreach (var child in node.Children)
        {
            if (child.ViaRelationship is { } relationshipId)
            {
                var relationship = _policy.GetRelationshipAccess(
                    node.EntityId, relationshipId, AuthorizationOperation.Read);

                if (!relationship.IsAllowed)
                    continue;

                var childAuthorized = AuthorizeNode(child, isRoot: false);
                if (childAuthorized is null)
                    continue;

                var combined = AuthorizationDecision.Combine(
                    relationship,
                    childAuthorized.Authorization is null
                        ? AuthorizationDecision.Allowed
                        : AuthorizationDecision.Conditional(childAuthorized.Authorization));

                children.Add(childAuthorized with
                {
                    Authorization = combined.Predicate
                });
                continue;
            }

            var authorizedChild = AuthorizeNode(child, isRoot: false);
            if (authorizedChild is not null)
                children.Add(authorizedChild);
        }

        return node with
        {
            Fields = fields.ToArray(),
            Children = children,
            Authorization = effective.Predicate
        };
    }

    private static AuthorizationDecision AuthorizationDecisionFromPredicate(AuthorizationPredicate? predicate) =>
        predicate is null
            ? AuthorizationDecision.Allowed
            : AuthorizationDecision.Conditional(predicate);
}